luaui/Widgets/dbg_test_runner.lua

```diff
@@ -221,9 +177,126 @@ local testRunState
 local activeTestState
 local resumeState
 local returnState
-local callinState
+local callinState = {callins = {}, recording = {}, unsafe = false}
 local spyControls
 
+
+-- callin tracking
+-- =========
+
+-- Hook callin
+-- Will register to count either predicate success or execution counts.
+-- predicate will be passed the callin arguments.
+-- @string name Callin name
+-- @func predicate Test function or false to just count callin calls
+-- @param target Object registering the callin, default: test_runner widget
+-- @number depth stack depth (normally for internal use)
+function registerCallin(name, predicate, target, depth)
+	local REGISTER_COUNT = 0
+	local REGISTER_FULL = 1
+
+	local depth = depth + 1
+	if not target then
+		target = widget
+	end
+	if not callinState.unsafe and not callinState.recording[name] then
+		error("[registerCallin:" .. name .. "] need to call expectCallin(\"" .. name .. "\") first", depth)
+	end
+	local mode = predicate and REGISTER_FULL or REGISTER_COUNT
+	if not callinState.callins[name] then
+		callinState.buffer[name] = {}
+	        callinState.counts[name] = 0
+	elseif callinState.callins[name] == REGISTER_COUNT and callinState.callins[name] ~= mode then
+		error("[registerCallin:" .. name .. "] expecting countOnly but requesting full", depth)
+	end
+	local countFunc
+	local counts = callinState.counts
+	if mode == REGISTER_COUNT and callinState.callins[name] == REGISTER_FULL then
+		counts[name] = #callinState.buffer[name]
+		callinState.buffer[name] = {}
+	end
+	if predicate then
+		countFunc = function(_, ...)
+			local res = predicate(...)
+			if res then
+				counts[name] = counts[name] + 1
+			end
+		end
+		for _, args in pairs(callinState.buffer[name]) do
+			countFunc(target, unpack(args))
+		end
+		callinState.buffer[name] = {}
+	else
+		countFunc = function()
+			counts[name] = counts[name] + 1
+		end
+	end
+	target[name] = countFunc
+	widgetHandler:UpdateWidgetCallInRaw(name, widget)
+	callinState.callins[name] = mode
+end
+
+-- Pre-Hook callin
+-- Will start prerecording so registerCallin will have access to previous callin executions
+-- @string name Callin name
+-- @func full Buffer all call arguments when true or just count number of executions
+-- @param target Object registering the callin, default: test_runner widget
+-- @number depth stack depth (normally for internal use)
+function preRegisterCallin(name, full, target, depth)
+	local depth = depth + 1
+	if callinState.recording[name] then
+		error("[preRegisterCallin:" ..  name .. "] already pre-registered", depth)
+	elseif callinState.callins[name] then
+		error("[preRegisterCallin:" .. name .. "] already registered", depth)
+	end
+	local predicate = false
+	if full then
+		local buffers = callinState.buffer
+		predicate = function(...)
+			local buffer = buffers[name]
+			local args = {...}
+			buffer[#buffer+1] = args
```

Comment:
This doesn't look like a predicate to me.

Reply:
Well, actually maybe the terminology doesn't match exactly, since I'm changing the root predicates to just count, and then the user provided predicate is passed to the one that's doing the counting. This one is passed into the same place, so even if terminology doesn't exactly match, I thought it's better to name it like this to follow where it goes easier.

That said, I can still call it func, tbh it was called func originally in my code, but I changed it to `predicate` since, as I said, thought it would be clearer.

Reply:
Ok, I understand your concerns. I changed the name, see if you like that better, otherwise don't hesitate to come back for more :).

Regarding how it works, I think you got it right, we're first registering a callin that just accumulates data, later on we do register a different one that actually does predicate work, and all previously collected data run through the real predicate.

Reply:
I made an example commit with minimal changes that would do away with the overloading: https://github.com/beyond-all-reason/Beyond-All-Reason/commit/2ef72086a90f7a800174bdc88fc5cba70966268e

I added a second commit where I removed an optimization: https://github.com/beyond-all-reason/Beyond-All-Reason/commit/656d72e4b01084c58d1bafa4388f0833015f6e5d

I personally value code maintainability (which usually equals readability) over performance and doubly so when considering tests. But I understand that you like to optimize for performance where possible. The commits I made are just suggestions, feel free to ignore or implement in your own way as you see fit. Even if you were to ignore my suggestions, the code is already a lot better than it was in the first version.

Reply:
All right, thx for the code!

Based on that I made two different branches so you can see and chose what you prefer, or give further feedback on any of them:

- [reorg](https://github.com/saurtron/Beyond-All-Reason/blob/test-runner-callin-improvements-reorg/luaui/Widgets/dbg_test_runner.lua#L184) - same as before but reorganized following (freely) your suggestions
- [stripped](https://github.com/saurtron/Beyond-All-Reason/blob/test-runner-callin-improvements-stripped/luaui/Widgets/dbg_test_runner.lua#L184) - reorganize a bit and strip counting functionality

Tbh, I thought the counting functionality can be useful in some cases (when needing to save on memory).

But I have no real usecase for that right now (well, I kind of do, since with stress test scenarios I may want to wait until a set of callins are done but really need to avoid all memory waste, but I can workaround generally by just waiting a fixed number of frames instead... it's [what I do atm](https://github.com/beyond-all-reason/Beyond-All-Reason/pull/4000/files#diff-944a60457f7c8524394de7cfcd61a00b2e0eea59997389b6b0bbcdbd5b2bca78R129) anyways). If it becomes an issue we can always revisit later.

Also, the recording is not compulsory here... that may help too, so no problem if you prefer to remove the "just counting" functionality in the name of readability, I'm all for that.

Reply:
I checked both versions and they both look good to me. I'm indifferent which one is chosen.

Thanks for all the hard work you are doing :+1:

Reply:
Ok, after giving it much though, and since you said you don't mind, I commited the reorg branch changes.

This way for cases where unsafe is used and only waitForCallin is used without a predicate it'll automatically be more efficient.

For other cases where just count is needed Test.expectCallin can also be used , although I imagine that won't be used as much since it requires awareness and manually passing the extra parameter.

Also, just added a 'count' parameter to waitForCallin and waitForCallinArgs, so any number of executions can be waited for, instead of just 1.